In [1]:
#import libraries
import pandas as pd
import numpy as np
import os

In [2]:
#creating a path
path = r'C:\Users\JocGa\Instacart Basket Analysis'

In [3]:
#import dataset
df_orders_products_merged = pd.read_pickle(os.path.join(path,'02 Data', 'Prepared Data', 'orders_products_merged.pkl'))

In [4]:
#overview of dataframe 
df_orders_products_merged.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                     NaN         196                  1          0   both   
1                    15.0         196                  1          1   both   
2                    21.0         196                  1          1   both   
3                    29.0         196                  1          1   both   
4                    28.0         196                  1          1   both   

  product_name  aisle_id  department_id  prices  
0         Soda        77              7     9.0  
1         Soda        77              7     9.0  
2         Soda        77              7     9.0  
3         Soda        77              7     9.0  
4         Soda        77              7     9.0

In [5]:
# creating a subset
df = df_orders_products_merged[:1000000]

# If-statement with User-Defined function

In [6]:
#checking subset was apply correctly
df.shape

(1000000, 14)

In [7]:
# Define a function
def price_label(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range'
  else: return 'Not enough data'

#The code above is the complete code for every condition you want to include for my filters.

In [8]:
# apply the function
df['price_range'] = df.apply(price_label, axis=1)

C:\Users\JocGa\AppData\Local\Temp\ipykernel_7644\1199080080.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


#the code above is creating new column with name 'price_range' and apply the price_label function on axis=1; red highlight simply informing you that there could be something problematic about the way you executed your procedure. 

In [9]:
df['price_range'].value_counts(dropna = False)

Mid-range product    756450
Low-range product    243550
Name: price_range, dtype: int64

#checking the values in new column; most of the prices land in the mid-range; note not showing 'high range' due to working on subset dataframe

In [10]:
#expensive product within the subset
df['prices'].max()

14.8

#the max of prices is 14.8; which the explain the mid-range product counts

# If-statements with the loc() Function

In [11]:
df_orders_products_merged.loc[df_orders_products_merged['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [12]:
df_orders_products_merged.loc[(df_orders_products_merged['prices'] <= 15) & (df_orders_products_merged['prices'] > 5), 'price_range_loc'] = 'Mid-range product' 

In [13]:
df_orders_products_merged.loc[df_orders_products_merged['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [14]:
df_orders_products_merged['price_range_loc'].value_counts(dropna = False)

Mid-range product     21861558
Low-range product     10130750
High-range product      412551
Name: price_range_loc, dtype: int64

#result on the whole dataframe; the price range for most products are in the mid-range

# If-statement with For-Loops

In [15]:
#days most order take place
df_orders_products_merged['orders_day_of_week'].value_counts(dropna = False)

0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: orders_day_of_week, dtype: int64

#data dictionary numbers are set to days of the week 0(sat) -6(fri); 0 means Saturday while 4 is wednesday 

In [16]:
#creating for-loop for orders_day_of_week
result = []

for value in df_orders_products_merged["orders_day_of_week"]:
    if value == 0:
        result.append("Busiest day")
    elif value == 4:
        result.append("Least busy")
    else:
        result.append("Regularly busy")

In [17]:
#print result
result

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Reg

In [18]:
#create new column within dataframe and set it equal to result
df_orders_products_merged['busiest_day'] = result

In [19]:
#print frequency
df_orders_products_merged['busiest_day'].value_counts(dropna = False)

Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: busiest_day, dtype: int64

#the days most order take place are on regularly busy which are on saturday and Wednesday is the least busy.

In [20]:
df_orders_products_merged

order_id  user_id  order_number  orders_day_of_week  \
0          2539329        1             1                   2   
1          2398795        1             2                   3   
2           473747        1             3                   3   
3          2254736        1             4                   4   
4           431534        1             5                   4   
...            ...      ...           ...                 ...   
32404854   1320836   202557            17                   2   
32404855     31526   202557            18                   5   
32404856    758936   203436             1                   2   
32404857   2745165   203436             2                   3   
32404858   3093936   205420             1                   4   

          order_hour_of_day  days_since_prior_order  product_id  \
0                         8                     NaN         196   
1                         7                    15.0         196   
2                        12                    21.0         196   
3                         7                    29.0         196   
4                        15                    28.0         196   
...                     ...                     ...         ...   
32404854                 15                     1.0       43553   
32404855                 11                     3.0       43553   
32404856                  7                     NaN       42338   
32404857                  5                    15.0       42338   
32404858                 14                     NaN       28818   

          add_to_cart_order  reordered _merge                   product_name  \
0                         1          0   both                           Soda   
1                         1          1   both                           Soda   
2                         1          1   both                           Soda   
3                         1          1   both                           Soda   
4                         1          1   both                           Soda   
...                     ...        ...    ...                            ...   
32404854                  2          1   both            Orange Energy Shots   
32404855                  2          1   both            Orange Energy Shots   
32404856                  4          0   both          Zucchini Chips, Pesto   
32404857                 16          1   both          Zucchini Chips, Pesto   
32404858                  8          0   both  Hot Oatmeal Multigrain Raisin   

          aisle_id  department_id  prices    price_range_loc     busiest_day  
0               77              7     9.0  Mid-range product  Regularly busy  
1               77              7     9.0  Mid-range product  Regularly busy  
2               77              7     9.0  Mid-range product  Regularly busy  
3               77              7     9.0  Mid-range product      Least busy  
4               77              7     9.0  Mid-range product      Least busy  
...            ...            ...     ...                ...             ...  
32404854        64              7     3.7  Low-range product  Regularly busy  
32404855        64              7     3.7  Low-range product  Regularly busy  
32404856        50             19     6.9  Mid-range product  Regularly busy  
32404857        50             19     6.9  Mid-range product  Regularly busy  
32404858       130             14    10.3  Mid-range product      Least busy  

[32404859 rows x 16 columns]

# Task

In [21]:
#days most order take place
df_orders_products_merged['orders_day_of_week'].value_counts(dropna = False)

0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: orders_day_of_week, dtype: int64

#data dictionary numbers are set to days of the week 0(Sat) -6(Fri)

In [22]:
# Creating the vector where the outputs from the loop will be stored
result2 = []

In [23]:
#creating for-loop for orders_day_of_week

for value in df_orders_products_merged["orders_day_of_week"]:
    if value == 0 or value == 1:
        result2.append("Busiest days")
    elif value == 4 or value == 3:
        result2.append("Slowest days")
    else:
        result2.append("Regularly busys")

In [24]:
# Create a new column in 'df_orders_products_merged' named busiest_days
df_orders_products_merged['busiest_days'] = result2

In [25]:
#overview of dataframe
df_orders_products_merged.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                     NaN         196                  1          0   both   
1                    15.0         196                  1          1   both   
2                    21.0         196                  1          1   both   
3                    29.0         196                  1          1   both   
4                    28.0         196                  1          1   both   

  product_name  aisle_id  department_id  prices    price_range_loc  \
0         Soda        77              7     9.0  Mid-range product   
1         Soda        77              7     9.0  Mid-range product   
2         Soda        77              7     9.0  Mid-range product   
3         Soda        77              7     9.0  Mid-range product   
4         Soda        77              7     9.0  Mid-range product   

      busiest_day     busiest_days  
0  Regularly busy  Regularly busys  
1  Regularly busy     Slowest days  
2  Regularly busy     Slowest days  
3      Least busy     Slowest days  
4      Least busy     Slowest days

In [26]:
#print frequency
df_orders_products_merged['busiest_days'].value_counts(dropna = False)

Regularly busys    12916111
Busiest days       11864412
Slowest days        7624336
Name: busiest_days, dtype: int64

#regularly busy has the most count such days like Mon, thrus, and Fri, Sat and Sun are the busiest days, and tue and wed are the slowest days

In [27]:
#column and rows counts
df_orders_products_merged.shape

(32404859, 17)

In [28]:
# Compute the order counts per hour
hourly_order_counts = df_orders_products_merged.groupby('order_hour_of_day')['order_id'].count()

In [29]:
# Compute the quartiles of the order counts
quartiles = np.quantile(hourly_order_counts, [0.25, 0.5, 0.75])

In [30]:
# Define a function to label each hour based on its order count
def label_hour(row):
    if row['order_count'] <= quartiles[0]:
        return 'Fewest orders'
    elif row['order_count'] <= quartiles[1]:
        return 'Average orders'
    else:
        return 'Most orders'

In [31]:
# Create a new column containing the labels
hourly_order_counts_df = hourly_order_counts.reset_index().rename(columns={'order_id': 'order_count'})
hourly_order_counts_df['busiest_period_of_day'] = hourly_order_counts_df.apply(label_hour, axis=1)

In [32]:
# Merge the labels back into the original dataframe
df_orders_products_merged = pd.merge(df_orders_products_merged, hourly_order_counts_df[['order_hour_of_day', 'busiest_period_of_day']], on='order_hour_of_day', how='left')

In [33]:
#overview of dataframe
df_orders_products_merged.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                     NaN         196                  1          0   both   
1                    15.0         196                  1          1   both   
2                    21.0         196                  1          1   both   
3                    29.0         196                  1          1   both   
4                    28.0         196                  1          1   both   

  product_name  aisle_id  department_id  prices    price_range_loc  \
0         Soda        77              7     9.0  Mid-range product   
1         Soda        77              7     9.0  Mid-range product   
2         Soda        77              7     9.0  Mid-range product   
3         Soda        77              7     9.0  Mid-range product   
4         Soda        77              7     9.0  Mid-range product   

      busiest_day     busiest_days busiest_period_of_day  
0  Regularly busy  Regularly busys           Most orders  
1  Regularly busy     Slowest days        Average orders  
2  Regularly busy     Slowest days           Most orders  
3      Least busy     Slowest days        Average orders  
4      Least busy     Slowest days           Most orders

In [34]:
#new number of columns and rows
df_orders_products_merged.shape

(32404859, 18)

In [35]:
df_orders_products_merged['busiest_period_of_day'].value_counts()

Most orders       27818650
Average orders     3989881
Fewest orders       596328
Name: busiest_period_of_day, dtype: int64

#the most orders lies on the 75% of values, any value that is less than or equal to quartiles[0] is in the bottom 25% of order counts and would be labeled as "Fewest orders". any value that is less than or equal to quartiles[1] is in the bottom 50% of order counts and would be labeled as "Average orders"

# Export Dataframe 

In [36]:
#export dataframe as a pickle file
df_orders_products_merged.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged_update.pkl'))